installers/Config.hs

```diff
@@ -0,0 +1,134 @@
+optReadLower = opt (diagReadCaseInsensitive . unpack)
+argReadLower :: (Bounded a, Enum a, Read a, Show a) => ArgName -> Optional HelpMessage -> Parser a
+argReadLower = arg (diagReadCaseInsensitive . unpack)
+
+data Options = Options
+  { oAPI           :: API
+  , oBuildJob      :: Maybe BuildJob
+  , oCertPass      :: Maybe Text
+  , oCluster       :: Cluster
+  , oDaedalusVer   :: Version
+  , oOutput        :: Text
+  , oPullReq       :: Maybe PullReq
+  , oTestInstaller :: TestInstaller
+  , oCI            :: CI
+  }
+
+optionsParser :: Parser Options
+optionsParser = Options
+  <$> (fromMaybe Cardano <$> (optional $
+                optReadLower "api"                 'a' "Backend API:  cardano or etc"))
+  <*> (optional   $
+      (BuildJob  <$> optText "build-job"           'b' "CI Build Job/ID"))
+  <*> (optional   $
+                     optText "cert-pass"           'p' "Certificate password")
+  <*> (fromMaybe Mainnet <$> (optional $
+                optReadLower "cluster"             'c' "Cluster the resulting installer will target:  mainnet or staging"))
+  <*> (fromMaybe "dev"   <$> (optional $
+      (Version   <$> optText "daedalus-version"    'v' "Daedalus version string")))
+  <*>                optText "output"              'o' "Installer output file"
+  <*> (optional   $
+      (PullReq   <$> optText "pull-request"        'r' "Pull request #"))
+  <*> flag TestInstaller     "test-installer"      't' "Test installers after building"
+  <*> pure Buildkite -- NOTE: this is filled in by auto-detection
+
+
+
+lshow :: Show a => a -> Text
+lshow = pack . fmap Data.Char.toLower . show
+
+dhallTopExpr :: Text -> Config -> OS -> Cluster -> Text
+dhallTopExpr path Launcher os cluster = path <> "/launcher.dhall ( "<>path<>"/" <> lshow cluster <> ".dhall "<>path<>"/" <> lshow os <> ".dhall ) "<>path<>"/" <> lshow os <> ".dhall"
+dhallTopExpr path Topology os cluster = path <> "/topology.dhall ( "<>path<>"/" <> lshow cluster <> ".dhall "<>path<>"/" <> lshow os <> ".dhall )"
+
+generateConfig :: Request -> FilePath -> FilePath -> IO ()
+generateConfig Request{..} configRoot outFile = handle $ do
+  GHC.IO.Encoding.setLocaleEncoding GHC.IO.Encoding.utf8
+
+  Dhall.detailed $ do
+    let inText = LT.fromStrict $ dhallTopExpr (pack configRoot) rConfig rOS rCluster
+    print inText
+    expr <- case Dhall.Parser.exprFromText (Directed "(stdin)" 0 0 0 0) inText of
+              Left  err  -> Control.Exception.throwIO err
+              Right expr -> return expr
+
+    expr' <- Dhall.Import.load expr
+    case Dhall.TypeCheck.typeOf expr' of
+      Left  err -> Control.Exception.throwIO err
+      Right _   -> return ()
+
+    json <- case Dhall.JSON.dhallToJSON expr' of
+              Left err  -> Control.Exception.throwIO err
+              Right json -> return json
+
+    Data.ByteString.writeFile outFile (Data.Yaml.encode json)
+
+handle :: IO a -> IO a
```

Comment:
Could you add a comment here why this was introduced?

Reply:
Done.